Adds management command for training ML

lfaucon opened the pull request.

AlbericC commented on ml/management/commands/ml_train.py:
```diff
@@ -0,0 +1,210 @@
+        [ccs.comparison.user_id, ccs.comparison.video_1_id, 
+            ccs.comparison.video_2_id, ccs.criteria, ccs.score, ccs.weight]
+        for ccs 
+        in ComparisonCriteriaScore.objects.all().prefetch_related("comparison")]
```
protect this interpreted list in case ccs does not hold a `comparison` element or if (due to a mishap) any other element is missing

OscarVlld replied:
@lfaucon

lfaucon replied:
These fields are supposed to be required. I would prefer not adding the fix that you propose, because if a mishap happens that would allow us to quickly discover it and patch it instead of failing silently

AlbericC commented on ml/management/commands/ml_train.py:
```diff
@@ -0,0 +1,210 @@
+
+    comparison_data: output of fetch_data()
+    criteria: str, rating criteria
+    
```
```suggestion
    resume: bool, resume previous processing
```

AlbericC commented on ml/management/commands/ml_train.py:
```diff
@@ -0,0 +1,210 @@
+    }
+    ratings_to_create = set(
+        (contributor_id, video_id)
+        for contributor_id, video_id, criteria_name, score, uncertainty 
```
If you don't use all elements in a tuple, you may as well use the dummy variable `_` to explicitely throw away the creteria_name, score, uncertainty 

or, use a variable to hold the full tuple and use `tup[0]` and `tup[1]`, so this statement will be valid even with incomplete tuples

OscarVlld replied:
@lfaucon

AlbericC reviewed (commented):
This code is fine by me, but it consists of critical low-level functions : some testing functions (for pytest) would be really useful to ensure everything works in all intended cases

OscarVlld commented on settings/settings.py:
```diff
@@ -108,7 +109,7 @@
         'PASSWORD': 'DATABASE_PASSWORD' in server_settings and server_settings['DATABASE_PASSWORD'] or '',
         'HOST': "DATABASE_HOST" in server_settings and server_settings["DATABASE_HOST"] or 'localhost',
-        'PORT': "DATABASE_PORT" in server_settings and server_settings["DATABASE_PORT"] or 5432,
+        'PORT': "DATABASE_PORT" in server_settings and server_settings["DATABASE_PORT"] or 9000,
```
```suggestion
        'PORT': "DATABASE_PORT" in server_settings and server_settings["DATABASE_PORT"] or 5432,
```

OscarVlld commented on settings/settings.py:
```diff
@@ -25,7 +25,7 @@
         server_settings = yaml.load(f, Loader=yaml.FullLoader)
 except FileNotFoundError:
-    print('No local settings.')
+    print('No local settings.--')
```
```suggestion
    print('No local settings.')
```

OscarVlld commented on ml/ml_requirements.txt:
```diff
@@ -0,0 +1,5 @@
+
+torch==1.9.0
+numpy==1.20.3
+matplotlib==3.4.2 
```
Remove matplotlib because not used in production ?

lfaucon replied:
No, let's keep it

OscarVlld commented on settings-tournesol.yaml:
```diff
@@ -0,0 +1,17 @@
+DEBUG: True
```
What about these 2 "settings-tournesol.yaml" ?

lfaucon replied:
I'll delete these files, settings file are supposed to be only your personnal files and not committed to the repository. The repository only contains a single example setting file

OscarVlld commented on ml/management/commands/plots.py:
```diff
@@ -0,0 +1,158 @@
+import matplotlib.pyplot as plt
```
Is it ok now ?

lfaucon replied:
It is better now, but I would still recommend as next step to completely separate the part used for experimenting with the ML algorithm. A good place for this would be the /scripts/ folder which contain useful pieces of code unnecessary for the backend to run.

OscarVlld commented on ml/management/commands/plots.py:
```diff
@@ -0,0 +1,158 @@
+    plt.ylabel(y)
+    plt.legend()
+
+# def clean_dic(dic):
```
Is it ok now ?

lfaucon replied:
Yes!

lfaucon commented:
## Changes

1. Moved most files to /ml/ from /ml/management/commands/ because in Django management commands are the scripts that you execute with `python manage.py <COMMAND_NAME>` and it is nice to keep this folder restricted to 1 file = 1 command
2. Use the list of criteria defined in the settings file. In general this sort of thing should be defined in a single place, we already had defined the list of criteria. If it is defined in multiple place it makes it harder over time to update it. Also the criteria “largely_recommended” was missing
3. Deleted the extra settings files



### Recommendations 

1. Write full variable names, not “crit” but “criteria”, not “glob” but “global”. One very important aspect of code is to be read as easily as possible by others. The contractions have the advantage to shorten the code, but I personally find that they make the code harder to understand


### Django 

The models (see /tournesol/models, Python classes importing django.models.Model) describe what the database contains. Each class corresponds to a database table and the attribute of the class to the columns of this table.

The rest of our backend will mostly be an API. It is composed of views, serializers and urls endpoints (see files with these names). They are used respectively for defining custom logic of our API, changing the data from our models when sending it to clients, defining the accessible urls.

You can experiment with the shell

`python manage.py shell`

Then run the code `from tournesol.models import ContributorRating`   Then `rating_list = ContributorRating.objects.all()`
